File: go/services/wallet_service.go
package services

import (
	"context"

	"github.com/trinsic-id/sdk/go/proto/services/universalwallet/v1/wallet"
)

// NewWalletService returns a wallet service with the base service configured
// using the provided options
func NewWalletService(options *Options) (WalletService, error) {
	base, err := NewServiceBase(options)
	if err != nil {
		return nil, err
	}
	service := &walletBase{
		Service: base,
		client:  wallet.NewUniversalWalletClient(base.GetChannel()),
	}

	return service, nil
}

// WalletService defines the interface for interacting with wallet
type WalletService interface {
	Service
	Search(userContext context.Context, request *wallet.SearchRequest) (*wallet.SearchResponse, error)
	InsertItem(userContext context.Context, request *wallet.InsertItemRequest) (*wallet.InsertItemResponse, error)
	DeleteItem(userContext context.Context, request *wallet.DeleteItemRequest) (*wallet.DeleteItemResponse, error)
}

type walletBase struct {
	Service
	client wallet.UniversalWalletClient
}

func (w *walletBase) Search(userContext context.Context, request *wallet.SearchRequest) (*wallet.SearchResponse, error) {
	if request == nil {
		request = &wallet.SearchRequest{}
	}

	if len(request.Query) == 0 {
		request.Query = "SELECT c.id, c.type, c.data FROM c OFFSET 0 LIMIT 100"
	}

	md, err := w.GetMetadataContext(userContext, request)
	if err != nil {
		return nil, err
	}

	response, err := w.client.Search(md, request)
	if err != nil {
		return nil, err
	}

	return response, nil
}

func (w *walletBase) InsertItem(userContext context.Context, request *wallet.InsertItemRequest) (*wallet.InsertItemResponse, error) {
	md, err := w.GetMetadataContext(userContext, request)
	if err != nil {
		return nil, err
	}

	response, err := w.client.InsertItem(md, request)
	if err != nil {
		return nil, err
	}

	return response, nil
}

func (w *walletBase) DeleteItem(userContext context.Context, request *wallet.DeleteItemRequest) (*wallet.DeleteItemResponse, error) {
	md, err := w.GetMetadataContext(userContext, request)
	if err != nil {
		return nil, err
	}

	response, err := w.client.DeleteItem(md, request)
	if err != nil {
		return nil, err
	}

	return response, nil
}

File: ruby/lib/services/wallet_service.rb
# frozen_string_literal: true

require 'services/service_base'

module Trinsic
  # Wallet Service wrapper
  class WalletService < ServiceBase
    def initialize(service_options = nil)
      super(service_options)
      if @service_options.server_use_tls
        channel_creds = GRPC::Core::ChannelCredentials.new
        @client = Wallet::UniversalWallet::Stub.new(url_string, channel_creds)
      else
        @client = Wallet::UniversalWallet::Stub.new(url_string, :this_channel_is_insecure)
      end
    end

    def search(request = nil)
      # request = Wallet::SearchRequest.new(query: query)
      request ||= Wallet::SearchRequest.new
      request.query = request.query.empty? ? 'SELECT c.id, c.type, c.data FROM c OFFSET 0 LIMIT 100' : request.query
      @client.search(request, metadata: metadata(request))
    end

    def insert_item(request)
      # request = Wallet::InsertItemRequest.new(item_json: JSON.generate(item))
      @client.insert_item(request, metadata: metadata(request))
    end

    def delete_item(request)
      @client.delete_item(request, metadata: metadata(request))
    end
  end
end
